File: ts/src/memo/issuer-tag.ts
import { keccak_256 } from "@noble/hashes/sha3";

/**
 * Derive an issuer tag from a namespace string.
 *
 * Computes `keccak256(namespace)` and takes the first 8 bytes as a `uint64` bigint.
 * Deterministic: the same namespace always produces the same tag.
 *
 * @param namespace - Application namespace (e.g. `"my-app"`, `"acme-payments"`)
 * @returns `uint64` bigint (8 bytes of keccak256 hash)
 * @example
 * ```ts
 * const tag = issuerTagFromNamespace("my-app");
 * const memo = encodeMemoV1({ type: "invoice", issuerTag: tag, ulid });
 * ```
 */
export function issuerTagFromNamespace(namespace: string): bigint {
  const encoded = new TextEncoder().encode(namespace);
  const hash = keccak_256(encoded);
  let tag = 0n;
  for (let i = 0; i < 8; i++) {
    tag = (tag << 8n) | BigInt(hash[i]!);
  }
  return tag;
}

File: rs/tempo-reconcile/src/memo/issuer_tag.rs
use sha3::{Digest, Keccak256};

/// Derive a deterministic issuer tag from a namespace string.
///
/// Algorithm: `first 8 bytes of keccak256(utf8(namespace))` as u64 big-endian.
///
/// The same namespace always produces the same tag. Tags prevent cross-application
/// memo collisions without requiring a central registry.
///
/// # Example
/// ```
/// use tempo_reconcile::issuer_tag_from_namespace;
/// let tag = issuer_tag_from_namespace("my-app");
/// ```
pub fn issuer_tag_from_namespace(namespace: &str) -> u64 {
    let hash = Keccak256::digest(namespace.as_bytes());
    // keccak256 always produces 32 bytes; extract first 8 with explicit indexing.
    u64::from_be_bytes([
        hash[0], hash[1], hash[2], hash[3], hash[4], hash[5], hash[6], hash[7],
    ])
}
